R/compat-obj-type.R

```diff
@@ -207,27 +210,57 @@ obj_type_oo <- function(x) {
 stop_input_type <- function(x,
                             what,
                             ...,
+                            allow_na = FALSE,
+                            allow_null = FALSE,
                             arg = caller_arg(x),
                             call = caller_env()) {
   # From compat-cli.R
-  format_arg <- env_get(
-    nm = "format_arg",
+  cli <- env_get_list(
+    nms = c("format_arg", "format_code"),
     last = topenv(),
-    default = NULL,
+    default = function(x) sprintf("`%s`", x),
     inherit = TRUE
   )
-  if (!is.function(format_arg)) {
-    format_arg <- function(x) sprintf("`%s`", x)
+
+  if (allow_na) {
+    what <- c(what, cli$format_code("NA"))
+  }
+  if (allow_null) {
+    what <- c(what, cli$format_code("NULL"))
+  }
+  if (length(what)) {
+    what <- chr_enumerate(what)
   }
 
   message <- sprintf(
     "%s must be %s, not %s.",
-    format_arg(arg),
+    cli$format_arg(arg),
     what,
     obj_type_friendly(x)
   )
 
   abort(message, ..., call = call, arg = arg)
 }
 
+chr_enumerate <- function(chr, sep = ", ", final = "or") {
```

Comment:
FWIW I'd call this function `oxford_comma` because that explains precisely what it does.